Machines/Acorn/Archimedes/Video.hpp

```diff
@@ -369,32 +248,61 @@ struct Video {
 	uint32_t cursor_shift_ = 0;
 	Timing horizontal_timing_, vertical_timing_;
 
+	enum class Depth {
+		OneBPP = 0b00,
+		TwoBPP = 0b01,
+		FourBPP = 0b10,
+		EightBPP = 0b11,
+	};
+
 	// Current video state.
 	enum class Phase {
 		Sync, Blank, Border, Display,
 	};
+	template <bool is_vertical>
 	struct State {
 		uint32_t position = 0;
+		uint32_t display_start = 0;
+		uint32_t display_end = 0;
 
 		void increment_position(const Timing &timing) {
-			++position;
-			if(position == 1024) position = 0;
+			if(position == timing.sync_width)		state |= SyncEnded;
+			if(position == timing.display_start)	{ state |= DisplayStarted; display_start = position; }
+			if(position == timing.display_end)		{ state |= DisplayEnded; display_end = position; }
+			if(position == timing.border_start)		state |= BorderStarted;
+			if(position == timing.border_end)		state |= BorderEnded;
+
+			cursor_active |= position == timing.cursor_start;
+			cursor_active &= position != timing.cursor_end;
 
 			if(position == timing.period) {
 				state = DidRestart;
 				position = 0;
+
+				// Both display start and end need to be seeded as bigger than can be reached,
+				// while having some overhead for addition.
+				display_end = display_start = std::numeric_limits<uint32_t>::max() >> 1;
+			} else {
+				++position;
+				if(position == 1024) position = 0;
 			}
+		}
 
-			if(position == timing.sync_width)		state |= SyncEnded;
-			if(position == timing.display_start)	state |= DisplayStarted;
-			if(position == timing.display_end)		state |= DisplayEnded;
-			if(position == timing.border_start)		state |= BorderStarted;
-			if(position == timing.border_end)		state |= BorderEnded;
+		bool is_outputting(Depth depth) const {
+			return position >= display_start + output_latencies[static_cast<uint32_t>(depth)] && position < display_end + output_latencies[static_cast<uint32_t>(depth)];
```

Comment:
Could this be replaced by a check on `state & DisplayStarted`?

Reply:
My answer when writing this would have been that DisplayStarted is locked in when the raster runs over that spot in the line, but the output latency may vary if colour depth changes during the line.

... but now I'm being told that all timing parameters — including colour depth — are locked in at the start of the field. If so then obviously this whole code can be a heck of a lot simpler, and probably not especially similar to its current form.